Add Sensors Setup Guide (#194)
ros-planning/navigation.ros.org: setup_guides/sensors/setup_sensors.rst

Common Sensor Messages
======================  

In this subsection, we discuss some of the common types of ``sensor_msgs`` you might encounter when setting up Nav2. We will provide a brief description for each sensor, an image of it being simulated in Gazebo and the corresponding visualization of the sensor readings in RViz.

.. note::  There are other types of ``sensor_msgs`` aside from the ones listed below.  The complete list of messages and their definitions can be found in the `sensor_msgs documentation <http://wiki.ros.org/sensor_msgs>`_.

sensor_msgs/LaserScan
---------------------

This message represents a single scan from a planar laser range-finder. This message is used in ``slam_toolbox`` and ``nav2_amcl`` for localization and mapping, or in ``nav2_costmap_2d`` for perception.

> .. image:: images/sensor_laserscan_gazebo.png

Reviewer: For all image pairs, can you get them to render in the middle of the page? These 2 are nearly 50% so they effectively so, but the next one for instance is off to the left noticeably.

    :width: 49 %

.. image:: images/sensor_laserscan_rviz.png
    :class: with-border
    :width: 49 %   

sensor_msgs/PointCloud2
-----------------------

This message holds a collection of 3D points, plus optional additional information about each point. This can be from a 3D lidar, a 2D lidar, a depth camera or more.

.. image:: images/sensor_pointcloud2_gazebo.png
    :width: 35 %

.. image:: images/sensor_pointcloud2_rviz.png
    :width: 49 %  

sensor_msgs/Range
-----------------

This is a single range reading from an active ranger that emits energy and reports one range reading that is valid along an arc at the distance measured. A sonar, IR sensor, or 1D range finder are examples of sensors that use this message.

.. image:: images/sensor_range_gazebo.png
    :width: 49 %

.. image:: images/sensor_range_rviz.png
    :width: 49 %   

sensor_msgs/Image
-----------------

This represents the sensor readings from RGB or depth camera, corresponding to RGB or range values.

.. image:: images/sensor_image_gazebo.png
    :width: 43 %

.. image:: images/sensor_image_rviz.png
    :width: 49 %   

Simulating Sensors using Gazebo
*******************************

> To give you a better grasp of how to set up sensors on a simulated robot, we will build up on our previous tutorials and attach sensors to our simulated robot ``sam_bot``. Similar to the previous tutorial where we used Gazebo plugins to add odometry sensors to ``sam_bot``, we will be using the Gazebo plugins to simulate a lidar sensor and a depth camera on ``sam_bot``. 

Reviewer: ```suggestion
To give you a better grasp of how to set up sensors on a simulated robot, we will build up on our previous tutorials and attach sensors to our simulated robot ``sam_bot``. Similar to the previous tutorial where we used Gazebo plugins to add odometry sensors to ``sam_bot``, we will be using the Gazebo plugins to simulate a lidar sensor and a depth camera on ``sam_bot``. If you are working with a real robot, most of these steps are still required for setting up your URDF frames and it will not hurt to also add in the gazebo plugins for later use. 
```

.. image:: images/add_topic_laserscan.png
    :align: center

Next, set the ``Reliability Policy`` to ``Best Effort`` and set the ``size`` to 0.1 to see the points clearer. You should see the visualized ``LaserScan`` detection as shown below. This corresponds to the detected cube and sphere that we added to the Gazebo world. 

.. image:: images/demo_laserscan_rviz.png
    :align: center

To visualize ``sensor_msgs/Image`` and ``sensor_msgs/PointCloud2``, do the same for topics ``demo/depth_camera/image_raw`` and ``/demo/depth_camera/points`` respectively:

.. image:: images/add_topic_image.png
    :width: 49 %

.. image:: images/add_topic_pointcloud2.png
    :width: 49 %

After adding the ``demo/depth_camera/image_raw`` topic in RViz, you should see the cube in the image window at the lower-left side of the RViz window, as shown below.

.. image:: images/demo_image_rviz.png
    :align: center

You should also see the ``sensor_msgs/PointCloud2``, as shown below.

.. image:: images/demo_pointcloud2_rviz.png
    :align: center

Mapping and Localization
************************

> Now that we have a robot with its sensors set up, we can use the obtained sensor information to build a map of the environment and to localize the robot on the map. The ``slam_toolbox`` package is a set of tools and capabilities for 2D Simultaneous Localization and Mapping (SLAM) in potentially massive maps with ROS2. It is also one of the officially supported SLAM library by Nav2, and we recommend to use this package in situations you need to use SLAM on your robot setup. Aside from the ``slam_toolbox``, localization can also be implemented through the ``nav2_amcl`` package. This package implements Adaptive Monte Carlo Localization (AMCL) which estimates the position and orientation of the robot in a map. Other techniques may also be available, please check Nav2 documentation for more information.

Reviewer: ```suggestion
Now that we have a robot with its sensors set up, we can use the obtained sensor information to build a map of the environment and to localize the robot on the map. The ``slam_toolbox`` package is a set of tools and capabilities for 2D Simultaneous Localization and Mapping (SLAM) in potentially massive maps with ROS2. It is also one of the officially supported SLAM libraries in Nav2, and we recommend to use this package in situations you need to use SLAM on your robot setup. Aside from the ``slam_toolbox``, localization can also be implemented through the ``nav2_amcl`` package. This package implements Adaptive Monte Carlo Localization (AMCL) which estimates the position and orientation of the robot in a map. Other techniques may also be available, please check Nav2 documentation for more information.
```

  | For the complete list of configuration parameters and example configuration of ``nav2_amcl``, see the `AMCL Configuration Guide <https://navigation.ros.org/configuration/packages/configuring-amcl.html>`_.

 
You can also refer to the `(SLAM) Navigating While Mapping guide <https://navigation.ros.org/tutorials/docs/navigation2_with_slam.html>`_ for the tutorial on how to use Nav2 with SLAM. You can verify that ``slam_toolbox`` and ``nav2_amcl`` have been correctly setup by visualizing the map and the robot's pose in RViz, similar to what was shown in the previous section.

Costmap_2D
**********
The costmap represents the environment of the robot in the form of an occupancy grid. The cells in the occupancy grid store cost values between 0-254 which denote a cost to travel through these zones. A cost of 0 means the cell is free while a cost of 254 means that the cell is lethally occupied. Values in between these extremes are used by navigation algorithms to steer your robot away from obstacles as a potential field. Costmaps in Nav2 are implemented through the ``nav2_costmap_2d`` package. 

> The costmap implementation consists of multiple layers, each of which has a certain function that contributes to a cell's overall cost. It consists of the following layers: static layer, inflation layer, range layer, obstacle layer, and voxel layer. The static layer represents the constant section of the costmap, like those produced by SLAM. The obstacle layer includes the obstacles reported by sensors in 2D. The voxel layer is similar to the obstacle layer but handles 3D data instead. The range layer allows for the inclusion of sonar and infrared data to our costmap. Lastly, the inflation layer represents the added cost values around lethal obstacles such that our robot avoids navigating close to obstacles. In the next subsection of this tutorial, we will have some discussion about the basic configuation of the different layers in ``nav2_costmap_2d``. 

Reviewer: ```suggestion
The costmap implementation consists of multiple layers, each of which has a certain function that contributes to a cell's overall cost. The package consists of the following layers, but are plugin-based to allow customization and new layers to be used as well: static layer, inflation layer, range layer, obstacle layer, and voxel layer. The static layer represents the map section of the costmap, like those produced by SLAM. The obstacle layer includes the obstacles reported by sensors in 2D. The voxel layer is similar to the obstacle layer but handles 3D data instead. The range layer allows for the inclusion of sonar and infrared data to our costmap. Lastly, the inflation layer represents the added cost values around lethal obstacles such that our robot avoids navigating into obstacles due to the robot's geometry. In the next subsection of this tutorial, we will have some discussion about the basic configuration of the different layers in ``nav2_costmap_2d``. 
```

The layers are integrated into the costmap through a plugin interface and then inflated using a user-specified `inflation radius <http://wiki.ros.org/costmap_2d/hydro/inflation>`_, if the inflation layer is enabled. For a deeper discussion on costmap concepts, you can have a look at the `ROS1 costmap_2D documentation <http://wiki.ros.org/costmap_2d>`_. Note that the ``nav2_costmap_2d`` package is mostly a straightforward ROS2 port of the ROS1 navigation stack version with minor changes required for ROS2 support and some new layer plugins.

Configuring Costmap_2D
======================

> In this subsection, we will show an example configuration of ``costmap_2d`` and explain the basic configuration parameters. The code block below is an example of ``costmap_2d`` configuration using static layer, obstacle layer, voxel layer, and inflation layer. Note that this block of parameters will be included in the `config/nav2_params.yaml <https://github.com/ros-planning/navigation2_tutorials/blob/master/sam_bot_description/config/nav2_params.yaml>`_ file  when we launch Nav2 at the end of this section.

Reviewer: ```suggestion
In this subsection, we will show an example configuration of ``costmap_2d`` and explain the basic configuration parameters for your sensors. The code block below is an example of ``costmap_2d`` configuration using static layer, obstacle layer, voxel layer, and inflation layer. Note that this block of parameters will be included in the `config/nav2_params.yaml <https://github.com/ros-planning/navigation2_tutorials/blob/master/sam_bot_description/config/nav2_params.yaml>`_ file  when we launch Nav2 at the end of this section.
```

.. code-block:: yaml
  :lineno-start: 1

  global_costmap:
    global_costmap:
      ros__parameters:
        update_frequency: 1.0
        publish_frequency: 1.0
        global_frame: map
        robot_base_frame: base_link
        use_sim_time: True
        robot_radius: 0.3
        resolution: 0.05
        track_unknown_space: true

>         rolling_window: true

Reviewer: This should be false for most situations, its only true generally for the local costmap

Build, Run and Verification
===========================

In this subsection, we will first configure ``slam_toolbox`` to make it subscribe to ``/demo/scan`` topic and then launch it to publish to ``/map`` topic and provide the ``map`` => ``odom`` transform. Recall that the ``map`` => ``odom`` transfrom is one of the primary requirements of the Nav2 system. The messages published on the ``/map`` topic will then be used by the static layer of the ``global_costmap``. 

Now that we have properly setup our robot descirption, odometry sensors, and necessary transforms, we can finally launch the Nav2 system itself. For now, we will only be exploring the costmap generation system of Nav2. After launching Nav2, we will visualize the costmaps in RViz to confirm our output. 

Launching slam_toolbox
----------------------

To be able to launch ``slam_toolbox``, make sure that you have installed the ``slam_toolbox`` package by executing the following command:

.. code-block:: shell

  sudo apt install ros-<ros2-distro>-slam-toolbox

> Now, we will configure the parameters of ``slam_toolbox``. Go to the ``config`` directory and create a ``mapper_params.yaml`` file. Then copy the contents of `config/mapper_params.yaml <https://github.com/ros-planning/navigation2_tutorials/blob/master/sam_bot_description/config/mapper_params.yaml>`_ and paste them into ``mapper_params.yaml``. 

Reviewer: So I think here just use my parameter files / launch files in SLAM Toolbox itself, not adding a new launch/config file in this package here (especially just for a demo unless you've remapped the scan topics to non-default `scan`). If you have, I think you should remove the `demo/` namespace since that isn't standard

Reviewer 2: We already removed the demo namespace and removed the new config file for slam toolbox.

Reviewer: They are still included in the tutorials PR

The important thing to note in the configuration of ``mapper_params.yaml`` is that we make ``slam_toolbox`` use the laser data published in ``/demo/scan`` by setting the ``scan_topic`` parameter to ``/demo/scan``. Again, the in-depth discussion of other parameters is not included in this tutorial, please refer to the official documentation of `slam_toolbox <https://github.com/SteveMacenski/slam_toolbox#readme>`_  for more information on other parameters. 

> Before we launch the ``slam_toolbox``, make sure you have launched sam_bot's ``display.launch.py`` by executing the the lines below. Remember that this provides our ``odom`` => ``base_link`` and ``base_link`` => ``sensors`` transformations. It also launches RViz which we can use to visualize the robot and Gazebo which acts as our physics simulator.

Reviewer: The URDF doesn't provide the odom->baselink transformation. Gazebo does through the diff drive plugin

Reviewer 2: The display.launch.py also launches Gazebo so the odom->baselink transform should be provided upon launching display.launch.py. We will make clarifications about this on the draft.

.. code-block:: shell

  colcon build
  . install/setup.bash
  ros2 launch sam_bot_description display.launch.py

Then, we will launch the ``async_slam_toolbox_node`` of ``slam_toolbox`` and set its parameters to ``mapper_params.yaml`` using the package's built-in launch files. Open a new terminal and then execute the following lines:

.. code-block:: shell

>   ros2 launch slam_toolbox online_async_launch.py slam_params_file:=<full/path/to/config/mapper_params.yaml>

Reviewer: Where is gazebo launched?

Reviewer 2: Gazebo is launched inside the display.launch.py

The ``slam_toolbox`` should now be publishing to the ``/map`` topic and providing the ``map`` => ``odom`` transform. 

We can verify in RViz that the ``/map`` topic is being published. In the RViz window, click the add button at the bottom-left part then go to ``By topic`` tab then select the ``Map`` under the ``/map`` topic. You should be able to visualize the message received in the ``/map`` as shown in the image below.

.. image:: images/map.png
    :align: center

We can also check that the transforms are correct by executing the following lines in a new terminal:

.. code-block:: shell

  ros2 run tf2_tools view_frames.py

The line above will create a ``frames.pdf`` file that shows the current transform tree. Your tranform tree should be similar to the one shown below:

.. image:: images/view_frames.png
    :align: center

Launching Nav2
--------------

Next, we will configure the costmap parameters of Nav2 before we launch it.

First, Make sure you have installed the Nav2 packages by executing the following:

.. code-block:: shell

  sudo apt install ros-<ros2-distro>-navigation2
  sudo apt install ros-<ros2-distro>-nav2-bringup

> We will configure the parameters of Nav2 to use our costmap configuration in the previous subsection. Create a file named ``nav2_params.yaml`` under the ``config`` directory and then copy the contents of `config/nav2_params.yaml <https://github.com/ros-planning/navigation2_tutorials/blob/master/sam_bot_description/config/nav2_params.yaml>`_ and paste them into the newly created ``nav2_params.yaml`` file.

Reviewer: As well here, if you remove the `demo` namespace, then you can just use the default parameters and launch files in Nav2 itself

Reviewer: Also, where is gazebo launched in this instruction?

Reviewer 2: The demo namespace and the new config file for nav2 have also been removed

Reviewer: OK, make sure to update the tutorials PR

For the other parameters of the obstacle layer, the ``max_obstacle_height`` parameter sets the maximum height of the sensor reading to return to the occupancy grid. The minimum height of the sensor reading can also be set using the ``min_obstacle_height`` parameter, which defaults to 0 since we did not set it in the configation. The ``clearing`` parameter is used to set whether the obstacle is to be removed from the costmap or not. The clearing operation is done by raytracing through the grid. The maximum and minimum range to raytrace clear objects from the costmap is set using the ``raytrace_max_range`` and ``raytrace_min_range`` respectively. The ``marking`` parameter is used to set whether the inserted obstacle is marked into the costmap or not. We also set the maximum and minimum range to mark obstacles in the costmap through the ``obstacle_max_range`` and ``obstacle_min_range`` respectively. 

For the inflation layer (lines 31-34 and 67-70), we set the exponential decay factor across the inflation radius using the ``cost_scaling_factor`` parameter. The value of the radius to inflate around lethal obstacles is defined using the ``inflation_radius``. 

> For the voxel layer (lines 51-66), we set the ``publish_voxel_map`` parameter to ``True`` to enable the publishing of the 3D voxel grid. The resolution of the voxels in height is defined using the ``z_resolution`` parameter, while the number of voxels in each column is defined using the ``z_voxels`` parameter. The ``mark_threshold`` parameter sets the minimum number of voxels in a column to mark as occupied in the occupancy grid. We set the ``observation_sources`` parameter of the voxel layer to ``scan``, and we set the scan parameters (in lines 61-66) similar to the parameters that we have discussed for the obstacle layer. As defined in its ``topic`` and ``data_type`` parameters, the voxel layer will use the ``LaserSCan`` published on the ``/scan`` topic by the lidar scanner.

Reviewer: ```suggestion
For the voxel layer (lines 51-66), we set the ``publish_voxel_map`` parameter to ``True`` to enable the publishing of the 3D voxel grid. The resolution of the voxels in height is defined using the ``z_resolution`` parameter, while the number of voxels in each column is defined using the ``z_voxels`` parameter. The ``mark_threshold`` parameter sets the minimum number of voxels in a column to mark as occupied in the occupancy grid. We set the ``observation_sources`` parameter of the voxel layer to ``scan``, and we set the scan parameters (in lines 61-66) similar to the parameters that we have discussed for the obstacle layer. As defined in its ``topic`` and ``data_type`` parameters, the voxel layer will use the ``LaserScan`` published on the ``/scan`` topic by the lidar scanner.
```
